contrib/epee/src/hex.cpp

```diff
@@ -88,22 +88,46 @@ namespace epee
   {
     std::vector<uint8_t> result;
 
-    // erase possible previous errors
-    errno = 0;
+    // the data to work with (std::string is always null-terminated)
+    auto data = src.data();
 
-    // iterate over all the input characters
-    for (size_t i = 0; i < src.size(); i += 2) {
-      // skip spacing and colons
-      if (std::isspace(src.data()[i]) || src.data()[i] == ':')
-        ++i;
+    // conver a single hex character to an unsigned integer
+    auto char_to_int = [](const char *input) {
+      switch (std::tolower(*input)) {
+        case '0': return  0;
+        case '1': return  1;
+        case '2': return  2;
+        case '3': return  3;
+        case '4': return  4;
+        case '5': return  5;
+        case '6': return  6;
+        case '7': return  7;
+        case '8': return  8;
+        case '9': return  9;
+        case 'a': return 10;
+        case 'b': return 11;
+        case 'c': return 12;
+        case 'd': return 13;
+        case 'e': return 14;
+        case 'f': return 15;
+        default:  throw std::range_error{ "Invalid hexadecimal input" };
+      }
+    };
 
-      // convert current character to number and add it to the vector
-      result.push_back(std::strtol(src.data() + i, nullptr, 16));
-    }
+    // keep going until we reach the end
+    while (data[0] != '\0') {
+      // skip spacing and colon
+      if (std::isspace(data[0]) || data[0] == ':') {
+        ++data;
+        continue;
+      }
 
-    // check for errors during conversion
-    if (errno)
-      throw std::range_error{ "Invalid hex-encoded characters detected" };
+      // convert two matching characters to int
+      auto high = char_to_int(data++);
+      auto low  = char_to_int(data++);
+
+      result.push_back(high << 4 | low);
+    }
```

Comment:
I'm curious, why this rather htan use the existing conversion code ?

Reply:
When I wrote this, it was because I hadn't seen the existing conversion code - seeing as it somehow wasn't included in this hex.h / hex.cpp

Later I found it is available in string_tools, but it doesn't support skipping spaces or colons. This is quite useful, as OpenSSL shows a certificates fingerprint separated with colons by default. I could of course also patch this function to ignore those characters.